pkg/ccl/changefeedccl/parquet_sink_cloudstorage.go

```diff
@@ -0,0 +1,289 @@
+// EncodeAndEmitRow which links the buffer in the cloudStorageSinkFile and the
+// parquetWriter every time we need to create a file for each unique combination
+// of topic and schema version (We need to have a unique parquetWriter for each
+// unique combination of topic and schema version because each parquet file can
+// have a single schema written inside it and each parquetWriter can only be
+// associated with a single schema.)
+type parquetCloudStorageSink struct {
+	wrapped *cloudStorageSink
+}
+
+type parquetFileWriter struct {
+	parquetWriter  *goparquet.FileWriter
+	schema         *parquetschema.SchemaDefinition
+	parquetColumns []pqexporter.ParquetColumn
+	numCols        int
+}
+
+func makeParquetCloudStorageSink(baseCloudStorageSink *cloudStorageSink) *parquetCloudStorageSink {
+	return &parquetCloudStorageSink{wrapped: baseCloudStorageSink}
+}
+
+// EmitRow does not do anything. It must not be called. It is present so that
+// parquetCloudStorageSink implements the Sink interface.
+func (parquetSink *parquetCloudStorageSink) EmitRow(
+	ctx context.Context,
+	topic TopicDescriptor,
+	key, value []byte,
+	updated, mvcc hlc.Timestamp,
+	alloc kvevent.Alloc,
+) error {
+	return errors.AssertionFailedf("Emit Row should not be called for parquet format")
+}
+
+// Close implements the Sink interface.
+func (parquetSink *parquetCloudStorageSink) Close() error {
+	return parquetSink.wrapped.Close()
+}
+
+// Dial implements the Sink interface.
+func (parquetSink *parquetCloudStorageSink) Dial() error {
+	return parquetSink.wrapped.Dial()
+}
+
+// EmitResolvedTimestamp does not do anything as of now. It is there to
+// implement Sink interface.
+func (parquetSink *parquetCloudStorageSink) EmitResolvedTimestamp(
+	ctx context.Context, encoder Encoder, resolved hlc.Timestamp,
+) error {
+	return errors.AssertionFailedf("Parquet format does not support emitting resolved timestamp")
+}
+
+// Flush implements the Sink interface.
+func (parquetSink *parquetCloudStorageSink) Flush(ctx context.Context) error {
+	return parquetSink.wrapped.Flush(ctx)
+}
+
+// EncodeAndEmitRow links the buffer in the cloud storage sync file and the
+// parquet writer (see parquetCloudStorageSink). It also takes care of encoding
+// and emitting row event to cloud storage. Implements the SinkWithEncoder
+// interface.
+func (parquetSink *parquetCloudStorageSink) EncodeAndEmitRow(
+	ctx context.Context,
+	updatedRow cdcevent.Row,
+	prevRow cdcevent.Row,
+	topic TopicDescriptor,
+	updated, mvcc hlc.Timestamp,
+	alloc kvevent.Alloc,
+) error {
+	s := parquetSink.wrapped
+	file, err := s.getOrCreateFile(topic, mvcc)
+	if err != nil {
+		return err
+	}
+	file.alloc.Merge(&alloc)
+
+	if file.parquetCodec == nil {
+		var err error
+		file.parquetCodec, err = makeParquetWriterWrapper(ctx, updatedRow, &file.buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	colOrd := -1
+	parquetRow := make(map[string]interface{}, file.parquetCodec.numCols)
```

Comment:
Please add a todo: to avoid map allocation on every call to emit row.
We actually need to allocate this map once per (schema) version.
These allocations are very expensive.